app/questionnaire/path_finder.py

```diff
@@ -37,51 +37,73 @@ def is_path_complete(self, path):
 
     def get_first_incomplete_location(self, path):
         if path:
-            first_location = path[0]
-            section_id = self.schema.get_section_id_for_block_id(
-                first_location.block_id
-            )
-
             for location in path:
                 block = self.schema.get_block(location.block_id)
                 block_type = block.get('type')
-                if location not in self.progress_store.get_completed_locations(
-                    section_id
-                ) and block_type not in ['SectionSummary', 'Summary', 'Confirmation']:
+
+                is_location_complete = (
+                    location
+                    in self.progress_store.get_completed_locations(
+                        section_id=location.section_id,
+                        list_item_id=location.list_item_id,
+                    )
+                )
+
+                if not is_location_complete and block_type not in [
+                    'SectionSummary',
+                    'Summary',
+                    'Confirmation',
+                ]:
                     return location
 
         return None
 
-    def full_routing_path(self):
+    def full_routing_path(self, sections=None):
         path = []
-        sections = self.schema.get_sections()
-        for section in sections:
-            path = path + list(self.routing_path(section))
+        schema_sections = sections or self.schema.get_sections()
+
+        for section_schema in schema_sections:
+            section_id = section_schema['id']
+            for_list = self.schema.get_repeating_list_for_section(section_id)
+
+            if for_list:
+                for list_item_id in self.list_store[for_list].items:
+                    path = path + list(
+                        self.routing_path(
+                            section_id=section_id, list_item_id=list_item_id
+                        )
+                    )
+            else:
+                path = path + list(self.routing_path(section_id=section_id))
+
         return path
 
-    def routing_path(self, section: Mapping) -> RoutingPath:
+    def routing_path(self, section_id, list_item_id=None) -> RoutingPath:
         """
         Visits all the blocks in a section and returns a path given a list of answers.
         """
         blocks: List[Mapping] = []
         path: List[Location] = []
+        section_schema = self.schema.get_section(section_id)
+        current_location = Location(section_id=section_id, list_item_id=list_item_id)
```

Comment:
A location object is needed since the location comparison in when rules got introduced.

Reply:
Reintroduced.